---
title: "Variant Filtering Script - R"
output:
  html_document:
    number_sections: no
    toc: yes
    css: style.css
---
Document generated at:
```{r, echo=FALSE}
Sys.time()
```
This script acts to accept and run filtering procedures on the generated variant, genotype and annotation tables generated by the bash portion of this filtering script.

This is a R markdown version of the R script component of the variant filtering script - it should not be used to run the script. Please see the R-script labelled variant_filtering.R on the MedGen servers in order to use this script.

The script requires its partner bash script and the variant_filtering.config file to run and for vcfs generated using GATK Unifiedgenotyper.


#Environment set up and Data import
Clear all existing data from environment (There should be none to begin with anyway!)
```{r, eval=TRUE, include=TRUE}
rm(list = ls())
```
Set working directory (args[1] applicable whilst running with bash script - else set manually) and import libraries
```{r, eval=FALSE, include=TRUE}
args = commandArgs(trailingOnly=TRUE)
setwd("V:/Philip/Bioinformatic Workflows and Reports/variant_filter")
```
```{r, eval=FALSE, include=TRUE}
require("stringr")
```
Several steps throughout the script log variant numbers to a `.log` file (Shown once)
```{r, eval=FALSE, include=TRUE}
###default variables & log start
clock <- as.character(Sys.time())
write(clock, file = "R_log.txt", append = FALSE)
write("##Variant Filter Script ## R-script Log - Log Begin", file = "R_log.txt", append = TRUE)
```
Import all data tables using `read.table` function & trim redundant columns:
```{r, eval=FALSE, include=TRUE}
###Import data tables from bash script
ad <- read.table("allelicdepth.table", header = TRUE,stringsAsFactors = FALSE)
anno <- read.table("annovar.table", header = TRUE, sep = "\t",stringsAsFactors = FALSE, quote="")
gq <- read.table("genoqual.table", header = TRUE,stringsAsFactors = FALSE)
gt <- read.table("genotype.table", header = TRUE,stringsAsFactors = FALSE)
dp <- read.table("sitedepth.table", header = TRUE,stringsAsFactors = FALSE)
config <- read.table("variant_filtering.config",stringsAsFactors = FALSE)
vv <- read.table("variant.table",comment.char = "",header = TRUE,stringsAsFactors = FALSE)
###remove columns that aren't needed
ad <- ad[,-(2:5)]
gq <- gq[,-(2:5)]
gt <- gt[,-(2:5)]
dp <- dp[,-(2:5)]
vv <- vv[,-(8:10)]
```
Column within the variant.table (imported as `vv`) renamed for safetey
```{r, eval=FALSE, include=TRUE}
names(vv)[4] <- "rsID"
```
Adding necessary annotation columns from `anno.table` output to variant file `vv`
```{r, eval=FALSE, include=TRUE}
###Add annotation cols to variant file
vv$GENE <- anno$Gene.refGene
vv$AA <- anno$AAChange.refGene
vv$CONSEQUENCE <- anno$ExonicFunc.refGene
vv$X1000G <- anno$X1000g2015aug_all
vv$EXAC <- anno$ExAC_ALL
vv$CADD <- anno$CADD_phred
```
#Variant Filtering Parameters and Calculations - Rarity & Prediction Algorithms
Loading specified parameters from .config file
```{r, eval=FALSE, include=TRUE}
###Rarity & damage filters
###import config file settings
X1000g <- config$V2[config$V1 == "1000G"]
exac <- config$V2[config$V1 == "EXAC"]
CADD <- config$V2[config$V1 == "CADD"]
HET <- config$V2[config$V1 == "HET"]
```
Variants selected that are lower than specified config values for both ExAC and 1000genomes
```{r, eval=FALSE, include=TRUE}
###filter variant ids that are below values for both 1000g & exac_all
rarity.ft <- as.data.frame(anno$ID[anno$X1000g2015aug_all < X1000g & anno$ExAC_ALL < exac])
names(rarity.ft)[1] <- "ID"
```
Variants selected that are above the specified config value or without a score for CADD variant damage prediction software
```{r, eval=FALSE, include=TRUE}
###filter variant ids that are above cadd
cadd.ft <- as.data.frame(anno$ID[anno$CADD_phred > CADD | anno$CADD_phred < 0])
names(cadd.ft)[1] <- "ID"
```
Merge filtered dataframes to find variants matching both criteria of previous filtering steps and apply selection to `vv`
```{r, eval=FALSE, include=TRUE}
###filter vv by rarity and cadd score
raritycadd.ft <- merge(rarity.ft, cadd.ft,sort = FALSE)
vv.ft <- vv[vv$ID %in% raritycadd.ft$ID,]
```
#Variant Filtering Parameters and Calculations - Functional Consequence & QUAL
Grep-style search in `annovar.table` to find functional consequences of interest - Stopgain, missense, etc. and apply filter to `vv` 
```{r, eval=FALSE, include=TRUE}
###filtering on functional consequnce - ExonicFunction.refGene
func.ft <- as.data.frame(anno$ID[grepl("nonsynonymous SNV",anno$ExonicFunc.refGene,fixed = TRUE)
                                 | grepl("stopgain",anno$ExonicFunc.refGene,fixed = TRUE)
                                 | grepl("stoploss",anno$ExonicFunc.refGene,fixed = TRUE)
                                 | grepl("frameshift insertion",anno$ExonicFunc.refGene,fixed = TRUE)
                                 | grepl("frameshift deletion",anno$ExonicFunc.refGene,fixed = TRUE)
                                 | grepl("nonframeshift insertion",anno$ExonicFunc.refGene,fixed = TRUE)
                                 | grepl("nonframeshift deletion",anno$ExonicFunc.refGene,fixed = TRUE)])

###rename col to match other ID cols
names(func.ft)[1] <- "ID"

###filter by functional consequence
vv.ft <- vv.ft[vv.ft$ID %in% func.ft$ID,]
```
Filter variant list to only those including a QUAL score of greater than 100 and apply filter to `vv`
```{r, eval=FALSE, include=TRUE}
###filter by qual
qual.ft <- as.data.frame(vv.ft$ID[vv.ft$QUAL > 100])
names(qual.ft)[1] <- "ID"
vv.ft <- vv.ft[vv.ft$ID %in% qual.ft$ID,]
###tidy variables
rm(rarity.ft,raritycadd.ft,func.ft,qual.ft)
```
#Calculation of local Hetrozygous and Homozygous rates

* Used as filters for in-batch variant rarity

* Expected that HET rate will be lower than 15-20% in all samples as it's unlikely to be the same causal variant

* Expected that HOM rate will always be lower than HET rate due to HWE

* Values are calculated from data in `genotype.table`

```{r, eval=FALSE, include=TRUE}
###Filter genotype information down to only retained variants after func, rarit & CADD filtering (limits CPU time)
gt <- gt[gt$ID %in% vv.ft$ID,]

###corce gt data.frame into a matrix and replace non-numeric format with numeric genotypes
###multi-allelic sites are replaced with -2 and missing sites with -9
gtm <- as.matrix(gt)
#additional handling of multiallelic sites
#gtm[!gtm == "0/0" | !gtm == "0/1" | !gtm == "1/1" | !gtm == "./."] <- -2
gtm[gtm == "0/0"] <- 0
gtm[gtm == "0/1"] <- 1
gtm[gtm == "1/1"] <- 2
gtm[gtm == "./."] <- -9
gt <- as.data.frame(gtm)
###apply function to sum the number of missing, het, hom and ref sites
refHOM <- apply(gtm,1, function(x) sum(x == 0))
HETp <- apply(gtm,1, function(x) sum(x == 1))
nonHOM <- apply(gtm, 1, function(x) sum(x == 2))
miss <- apply(gtm, 1, function(x) sum(x == -9))
###form matrix for pct calculations
calc <- cbind(refHOM,HETp,nonHOM,miss)

###calculate hetpct & hompct (excluding missing sites) and missingness over all sites
hetpct <- ((calc[,2] / (calc[,1] + calc[,2] + calc[,3]))*100)
hompct <- ((calc[,3] / (calc[,1] + calc[,2] + calc[,3]))*100)
misspct <- ((calc[,4] / length(gt[1,-1])*100))
###append values to new columns in vv.ft
vv.ft$HET_rate <- hetpct
vv.ft$HOM_rate <- hompct
vv.ft$MISS_rate <- misspct
```
As described previously, HET & HOM rate are used to filter variants and apply those filters to `vv` - Here HET rate must be < 15 and HOM rate cannot exceed HET rate for a particular site
```{r, eval=FALSE, include=TRUE}
###filter by het/hom ratio and het rate (het > hom & no het rate in cohort above 15%)
hethom.ft <- as.data.frame(vv.ft$ID[vv.ft$HET_rate < HET & vv.ft$HET_rate >= vv.ft$HOM_rate])
names(hethom.ft)[1] <- "ID" 

###Extract variants based on filtered list
vv.ft <- vv.ft[vv.ft$ID %in% hethom.ft$ID,]
###tidy variables and tables
rm(gtm,refHOM,HETp,nonHOM,miss,calc,hetpct,hompct,misspct,hethom.ft)
```
#Transformation of allelic depth/frequency table

* Performs a transformation of data in `allelicdepth.table`

* Generated a percentage calculation of reads supporting ALT call

* Used to filter based on sites where maximum ALT allelic depth is too low across all samples

* WARNING - Will filter out mosaic variants!
```{r, eval=FALSE, include=TRUE}
###performing allelic depth transformation to allele percent
###make copy of allelicdepth(ad)
af <- ad[ad$ID %in% vv.ft$ID,]

###loop over each sample column
for(i in 2:ncol(af)){
  ###split ad values on ,  to generate a maj allelic read depth and min allelic read depth
  ###replace empty cells in second col with "." to match first
  alfe <- as.data.frame(str_split(af[,i],",",simplify = TRUE),stringsAsFactors = FALSE)
  alfe$V2[alfe$V2 %in% ""] <- "."
  d <- data.frame(x=rep(0,nrow(af)))
  ###loop over new maj/min read depth dataframe & convert to float or replace with na
  for(r in 1:nrow(alfe)){
    if(alfe[r,2] == "."){d[r,1] <- NA}
    else{
      afmaj <- as.numeric(alfe[r,1])
      afmin <- as.numeric(alfe[r,2])
      aftotal <- afmaj + afmin
      afperct <- (afmin/aftotal)
      d[r,1] <- afperct}
  }
  
  ###replace column in af with new float/na values
  af[,i] <- as.data.frame(d)
  rm(d)
}

###tidy variables and tables
rm(alfe,afmaj,afmin,afperct,aftotal,i,r)
```
Performs the filtering on basis of maximum site ALT allelic depth on `vv` - in this case no higher than 0.30
```{r, eval=FALSE, include=TRUE}
###filter on variants with no af rate above threshold
af.ft <- data.frame(x=rep(0,nrow(af)))
for(i in 1:nrow(af)){
    ###Variants filtered on no variants af above than 0.3
    if(max(af[i,2:ncol(af)], na.rm = TRUE) > 0.3){
        af.ft[i,1] <- af[i,1]}
    else{
        af.ft[i,1] <- NaN
    }
}
names(af.ft)[1] <- "ID"
af.ft <- subset(af.ft, (af.ft[,1] != "NaN"))

vv.ft <- vv.ft[vv.ft$ID %in% af.ft$ID,]
rm(af.ft,i)
```
#Final output assembly - Genotype additions and writing output
On finalised list of filtered variants add all the genotype information for each sample
```{r, eval=FALSE, include=TRUE}
###Add genotype information for remaining variants
gt.ft <- gt[gt$ID %in% vv.ft$ID,]
vvgt <- merge(vv.ft,gt.ft, sort = FALSE)
```
Both the final `vv` table (named variant_filtering_results.txt) & corresponding allelic frequency table written by `write.table`
```{r, eval=FALSE, include=TRUE}
###rename ID col
names(vvgt)[1] <- "Id"
names(af)[1] <- "Id"
###write filtered table out
write.table(vvgt,file = "variant_filtering_results.tsv",sep = "\t",row.names = FALSE, quote = FALSE)
write.table(af,file = "variant_filtering_results_AF.table",sep = "\t",row.names = FALSE, quote = FALSE)
```

```{r Example output code, echo=FALSE, eval=TRUE, include=TRUE}
###update
#ID <- c("Var1", "Var24", "var76")
#CHROM <- c(3,5,10)
#POS <- c(234325,34859,12332)
#rsID <- c("rs14245","rs25562","rs686892")
#REF <- c("G","C","G")
#ALT <- c("T","A", "C")
#QUAL <- c(142,4339,755)
#GENE <- c("GENE1","GENE2","GENE3")
#CONSEQUENCE <- c("Missense","Stop_Gain","Missense")
#X1000g <- c(0.0004,-9,0.045)
#EXAC <- c(0.001,0.0043,0.002)
#CADD <- c(23,33,15)
#HET_rate <- c(3.5,5,2.4)
#HOM_rate <- c(0,1.2,0)
#Sample_1 <- c("0/1","0/0","0/0")
#Sample_2 <- c("0/0","0/0","0/1")
#exOUT <- data.frame(ID, CHROM, POS, rsID, REF, ALT, QUAL, GENE, CONSEQUENCE, X1000g, EXAC, CADD, HET_rate, HOM_rate, Sample_1, Sample_2)
library(knitr)
```
```{r, fig.align = "center", eval=TRUE, echo=FALSE}
#kable(exOUT, caption = "Example Output Table")
```
-------
<center>Philip Smith - Department of Medical Genetics, University of Cambridge - Variant Filtering Script</center>