---
title: "Variant Filtering Script - Bash"
output:
  html_document:
    css: style.css
    number_sections: no
    toc: yes
---
Document generated at:
```{r, echo=FALSE}
Sys.time()
```
This script acts to accept and run filtering procedures on the generated variant, genotype and annotation tables generated by the R portion of this filtering script.

This is a R markdown version of the bash script component of the variant filtering script - it should not be used to run the script. Please see the  bash script labelled variant_filtering.sh on the MedGen servers in order to use this script.

The script requires its partner R script and the variant_filtering.config file to run and for vcfs generated using GATK Unifiedgenotyper.
#Script Initialisation & Variable definitions
Defining the script as being .sh, defining non-command line static variables genome `BUILD` and annovar path `ANNO`, setting defaults for command-line variables 
```{bash, echo=TRUE, eval=FALSE}
#!/bin/bash
```
```{bash, echo=TRUE, eval=FALSE}
###Static variables
BUILD="hg38"
ANNO="/data/Resources/Software/"
```
```{bash, echo=TRUE, eval=FALSE}
###Command-line variables and script start
INPUT="NULL"
CONFIG="NULL"
PROJECT="Unknown_project"
TEMP="FALSE"
```
Proivded `--help` section and default output when no arguments are given
```{bash, echo=TRUE, eval=FALSE}
###Help block - provides details on arguments
for arg in "$@"; do
	if [[ "$arg" == "--help" ]]; then
		echo -e "
##Variant Filtering Script ## - Help Documentation
Options:
--project		Specifies a folder to generate results in - by default the folder is named 'Unknown_project_variantfiltering'.
			Folder is generated where the script is located - no output folder selection is implimented.
--Input			Full path to the VCF file you want to filter - REQUIRED
--config		Full path to the provided variant_filtering.config file provided with the script - should contain 2 column tab
			separated file with names and values for each filtering parameter. Current parameters are;
			
			NAME		TYPE		Effect
			MEANDP		Interger	Filters on mean read-depth across all samples provided on provided value
			MAF		Float		Filters sites in found in provided proportion of all samples in VCF
			GQ		Interger	Filters on genotype quality at a given site across all samples
			MISSING		Float		Filters sites in which the provided proportion of genotypes are missing
			1000G		Float		Filters in 1000G data for given rarity (R-Script AND logic with ExAC)
			EXAC		Float		Filters in ExAC data for given rarity (RScript - AND logic with 1000G)
			CADD		Interger	Filters CADD score for each site at the provided value
--TEMP			Only option TRUE - automatically retains the temporary .table files and asociated intermediate files when script
			is run using this option"			               
		exit
	fi
done

###Default output - No arguments
if [[ $# -eq 0 ]]; then
	echo -e `date +[%D-%R]` "## Variant Filter Script ## - You need to provide at least SOME arguments! Try using --help for documentation and examples!"
	exit
fi
```
Progress logging of `echo` commands throughout output to both stdout and a .log file via `tee` function
```{bash, echo=TRUE, eval=FALSE}
echo -e `date +[%D-%R]` "## Variant Filter Script ## - Script started" | tee -a variantfilter.log
```
Block required for parsing of arguments provided in the command-line for assignment to their corresponding variable
```{bash, echo=TRUE, eval=FALSE}
###Command-line argument passing
while [[ $# > 1 ]]
	do
	key="$1"
	case $key in
		--project)
        	echo -e `date +[%D-%R]` "## Variant Filter Script ## - Project name set to ${2}" | tee -a variantfilter.log
		PROJECT=${2}
		shift
		;;
		--input)
		echo -e `date +[%D-%R]` "## Variant Filter Script ## - Input VCF is ${2}" | tee -a variantfilter.log
		INPUT=${2}
		shift
		;;
		--config)
		echo -e `date +[%D-%R]` "## Variant Filter Script ## - Using config file ${2}" | tee -a variantfilter.log
		CONFIG=${2}
		shift
		;;
		--TEMP)
		echo -e `date +[%D-%R]` "## Variant Filter Script ## - Temporary file retention set to ${2}" | tee -a variantfilter.log
		R=${2}
		shift
		;;
	esac
	shift
done
```
Read over config file to find values provided for variables to be given to `vcftools` for vcf filtering
```{bash, echo=TRUE, eval=FALSE}
###config arugment settings
MEANDP=$(grep MEANDP ${CONFIG} | cut -f2)
echo -e `date +[%D-%R]` "## Variant Filter Script ## - Minimum Mean Read Depth set to ${MEANDP}" | tee -a variantfilter.log
MAF=$(grep MAF ${CONFIG} | cut -f2)
echo -e `date +[%D-%R]` "## Variant Filter Script ## - Maximum cohort Minor Allele Frequency set to ${MAF}" | tee -a variantfilter.log
GQ=$(grep GQ ${CONFIG} | cut -f2)
echo -e `date +[%D-%R]` "## Variant Filter Script ## - Minumum Genotype Quality set to ${GQ}" | tee -a variantfilter.log
MISSING=$(grep MISSING ${CONFIG} | cut -f2)
echo -e `date +[%D-%R]` "## Variant Filter Script ## - Maximum number of missing values at site to ${MISSING}" | tee -a variantfilter.log
```
Determine if a project folder has been generated, make it if not, and set working directory to that location
```{bash, echo=TRUE, eval=FALSE}
###Run folder setup and config file availability
if [[ ! -d ${PROJECT}_variantfiltering ]]; then
	mkdir ${PROJECT}_variantfiltering
else
	echo -e `date +[%D-%R]` "## Variant Filter Script ## - Project folder already exists - Overwritting Content" | tee -a variantfilter.log
fi
cp ${CONFIG} ${PROJECT}_variantfiltering/variant_filtering.config
cd ${PROJECT}_variantfiltering
```
#Variant Filtering - Basic filtering and Quality metrics
Using `vcftools` multi-sample vcf is filtered for the parameters provided in the .config file:

* minimum mean read depth over site

* maximum Minor Allele Frequency (local - aka across samples in vcf)

* minimum Genotype quality over site

* maximum percent of missing genotypes

Output is then recoded to a new vcf containing only sites matching all filtering criteria
```{bash, echo=TRUE, eval=FALSE}
####echo -en `date +[%D-%R]` "## Variant Filter Script ## - Filtering VCF on specified values..." | tee -a ../variantfilter.log
cp ${INPUT} variant_orig.vcf
vcftools --vcf variant_orig.vcf --min-meanDP ${MEANDP} --max-maf ${MAF} --minGQ ${GQ} --max-missing ${MISSING} --recode --out variant_orig > /dev/null 2>&1
mv variant_orig.recode.vcf variant_filtered.vcf
echo -e "\r"`date +[%D-%R]` "## Variant Filter Script ## - Filtering VCF on specified values...Done" | tee -a ../variantfilter.log
```
#Generation of .table files for R-script processing
A secondary vcf file containing only sites and no individuals is generated from the filtered vcf for use with Annovar which does not handle multi-sample vcfs - generates `variant.table`
```{bash, echo=TRUE, eval=FALSE}
###removing header command & generating intermediate files with bcftools
echo -en `date +[%D-%R]` "## Variant Filter Script ## - Generating intermediate files..." | tee -a ../variantfilter.log
vcftools --vcf variant_filtered.vcf --max-indv 0 --recode --out annotate > /dev/null 2>&1
sed -n '/#CHROM/,${p}' annotate.recode.vcf  > variant.table
echo -e "\r"`date +[%D-%R]` "## Variant Filter Script ## - Generating intermediate files...Done" | tee -a ../variantfilter.log
```
Using `bcftools` to extract genotype, read depth, and other quality metrics from the INFO field to generate the following .table files:

* genotype.table

* sitedepth.table

* allelicdepth.table

* genoqual.table

The `.table` files then are processed by several sed processes to make them compatible in downstream processes
```{bash, echo=TRUE, eval=FALSE}
echo -en `date +[%D-%R]` "## Variant Filter Script ## - Generating INFO field tables..." | tee -a ../variantfilter.log 
###Use bcftools to extract depth/genotype/INFO_field information
bcftools query --print-header -f '%CHROM\t%POS\t%REF\t%ALT[\t%GT]\n' -o genotype.table variant_filtered.vcf
sed -i 's/\[[0-9]\+\]//g' genotype.table
bcftools query --print-header -f '%CHROM\t%POS\t%REF\t%ALT[\t%DP]\n' -o sitedepth.table variant_filtered.vcf
sed -i 's/\[[0-9]\+\]//g' sitedepth.table
bcftools query --print-header -f '%CHROM\t%POS\t%REF\t%ALT[\t%AD]\n' -o allelicdepth.table variant_filtered.vcf
sed -i 's/\[[0-9]\+\]//g' allelicdepth.table
bcftools query --print-header -f '%CHROM\t%POS\t%REF\t%ALT[\t%GQ]\n' -o genoqual.table variant_filtered.vcf
sed -i 's/\[[0-9]\+\]//g' genoqual.table
###removing incorrect #CHROM to CHROM for R input
sed -i 's/# CHROM/CHROM/' genotype.table
sed -i 's/# CHROM/CHROM/' sitedepth.table
sed -i 's/# CHROM/CHROM/' allelicdepth.table
sed -i 's/# CHROM/CHROM/' genoqual.table
sed -i 's/#CHROM/CRHOM/' variant.table
###Removing bcftools tags
sed -i 's/:GT//g' genotype.table
sed -i 's/:DP//g' sitedepth.table
sed -i 's/:AD//g' allelicdepth.table
sed -i 's/:GQ//g' genoqual.table
echo -e "\r"`date +[%D-%R]` "## Variant Filter Script ## - Generating INFO field tables...Done" | tee -a ../variantfilter.log
```
Using the `variant.table`, annotation information regarding rarity, function, and damage predictions generated by Annovar for each site
```{bash, echo=TRUE, eval=FALSE}
###Generating annovar-annotation file for use as table - inc gMAF and damage predictions
echo -en `date +[%D-%R]` "## Variant Filter Script ## - Generating Annovar annotation file..." | tee -a ../variantfilter.log
${ANNO}annovar/convert2annovar.pl -format vcf4old annotate.recode.vcf --outfile annovarform > /dev/null 2>&1
${ANNO}annovar/table_annovar.pl annovarform ${ANNO}annovar/humandb/ -buildver ${BUILD} -out annotated -remove -protocol refGene,1000g2015aug_all,exac03,avsnp144,dbnsfp30a,clinvar_20160302,cosmic70,nci60,dbscsnv11 -operation g,f,f,f,f,f,f,f,f -nastring -9 > /dev/null 2>&1
mv annotated.hg38_multianno.txt annovar.table
echo -e "\r"`date +[%D-%R]` "## Variant Filter Script ## - Generating Annovar annotation file...Done" | tee -a ../variantfilter.log
```
#Table Quality control and Indexing
Confirmation of row equality between all tables so that all sites were processed and retained - report row lengths if unequal
```{bash, echo=TRUE, eval=FALSE}
###Check file length for the correct number of rows - all .tables & annovar files should have the same number of rows = to each vcf record
VAR=$(cat variant.table | wc -l)
GENOT=$(cat genotype.table | wc -l)
GENOQ=$(cat genoqual.table | wc -l)
ALLELIC=$(cat allelicdepth.table| wc -l)
SITE=$(cat sitedepth.table | wc -l)
ANNOTATE=$(cat annovar.table | wc -l)

if [ "$GENOT" -ne "$VAR" ] || [ "$GENOQ" -ne "$VAR" ] || [ "$ALLELIC" -ne "$VAR" ] || [ "$SITE" -ne "$VAR" ] || [ "$ANNOTATE" -ne "$VAR" ]; then
	echo -e `date +[%D-%R]` "## Variant Filter Script ## - Mismatched rows - The data tables had the following row counts:" | tee -a ../variantfilter.log
	echo -e `date +[%D-%R]` "Genotype.table - ${GENOT}" | tee -a ../variantfilter.log
        echo -e `date +[%D-%R]` "Genoqual.table - ${GENOQ}" | tee -a ../variantfilter.log
        echo -e `date +[%D-%R]` "Allelicdepth.table - ${ALLELIC}" | tee -a ../variantfilter.log
	echo -e `date +[%D-%R]` "Sitedepth.table - ${SITE}" | tee -a ../variantfilter.log
	echo -e `date +[%D-%R]` "Annovar.table - ${ANNOTATE}" | tee -a ../variantfilter.log
	echo -e `date +[%D-%R]` "Reference VCF - ${VAR}" | tee -a ../variantfilter.log
	echo -e `date +[%D-%R]` "## Variant Filter Script ## - Exiting now" | tee -a ../variantfilter.log
	exit
else
	echo -e `date +[%D-%R]` "## Variant Filter Script ## - All data tables contain matching rows" | tee -a ../variantfilter.log
fi
```
Given all tables are equal in length, generate an index column labeling each row from 1 to n rows with "Var n"
```{bash, echo=TRUE, eval=FALSE}
###index all the files with header ID followed by var1-var(nrows-1)
echo -en `date +[%D-%R]` "## Variant Filter Script ## - Indexing data tables..." | tee -a ../variantfilter.log
awk -F'\t' -v OFS='\t' 'NR == 1 {print "ID", $0; next} {print "Var"(NR-1), $0}' variant.table > awk.table
mv awk.table variant.table
awk -F'\t' -v OFS='\t' 'NR == 1 {print "ID", $0; next} {print "Var"(NR-1), $0}' genotype.table > awk.table
mv awk.table genotype.table
awk -F'\t' -v OFS='\t' 'NR == 1 {print "ID", $0; next} {print "Var"(NR-1), $0}' genoqual.table > awk.table
mv awk.table genoqual.table
awk -F'\t' -v OFS='\t' 'NR == 1 {print "ID", $0; next} {print "Var"(NR-1), $0}' allelicdepth.table > awk.table
mv awk.table allelicdepth.table
awk -F'\t' -v OFS='\t' 'NR == 1 {print "ID", $0; next} {print "Var"(NR-1), $0}' sitedepth.table > awk.table
mv awk.table sitedepth.table
awk -F'\t' -v OFS='\t' 'NR == 1 {print "ID", $0; next} {print "Var"(NR-1), $0}' annovar.table > awk.table
mv awk.table annovar.table
echo -e "\r"`date +[%D-%R]` "## Variant Filter Script ## - Indexing data tables...Done" | tee -a ../variantfilter.log
```
Assigning the working directory to a variable - for use in calling the R script
```{bash, echo=TRUE, eval=FALSE}
wd=`pwd`
```
Perform minor clean up of large intermediate files prior to R-script
```{bash, echo=TRUE, eval=FALSE}
###File clean up
rm annovarform
rm annotate.recode.vcf
```
#R-script command
Use `R-script` command to parse the variant_filtering.R script both the `.tables` and current working directory.
See R-markdown for the R component of this script for details of code and function
```{bash, echo=TRUE, eval=FALSE}
###Run R script to filter variants	
echo -e `date +[%D-%R]` "## Variant Filter Script ## - R Script Started" | tee -a ../variantfilter.log	
echo -en `date +[%D-%R]` "## Variant Filter Script ## - Completing filtering on consequence, allele frequency & rarity..." | tee -a ../variantfilter.log
Rscript ../variant_filtering.R ${wd} > /dev/null 2>&1
echo -e "\r"`date +[%D-%R]` "## Variant Filter Script ## - Completing filtering consequence, allele frequency & rarity...Done" | tee -a ../variantfilter.log
```
On completion of R script portion of the pipeline all temporary files are removed and log files moved to parent directory
```{bash, echo=TRUE, eval=FALSE}
###Clean up temporary files
if [ "$TEMP" == "FALSE" ]; then
	echo -en `date +[%D-%R]` "## Variant Filter Script ## - Removing temporary files..." | tee -a ../variantfilter.log
	mv ../variantfilter.log variantfilter.log
	rm allelicdepth.table
	rm annovar.table
	rm genoqual.table
	rm genotype.table
	rm sitedepth.table
	rm variant.table
	rm variant_orig.vcf
	rm variant_filtering.config
	echo -e "\r"`date +[%D-%R]` "## Variant Filter Script ## - Removing temporary files...Done" | tee -a variantfilter.log 
fi
echo -e `date +[%D-%R]` "## Variant Filter Script ## - Variant Filter Script Finsihed" | tee -a variantfilter.log
```

-------
<center>Philip Smith - Department of Medical Genetics, University of Cambridge - Variant Filtering Script</center>